packages/@aws-cdk/assets/lib/asset.ts

```diff
@@ -84,33 +86,45 @@ export class Asset extends cdk.Construct {
             description: `S3 bucket for asset "${this.path}"`,
         });
 
-        const keyParam = new cdk.Parameter(this, 'S3ObjectKey', {
+        this.s3PrefixParam = new cdk.Parameter(this, 'S3Prefix', {
+            type: 'String',
+            description: `S3 prefix for asset "${this.path}"`
+        });
+
+        const keyParam = new cdk.Parameter(this, 'S3VersionKey', {
             type: 'String',
-            description: `S3 object for asset "${this.path}"`
+            description: `S3 key for asset version "${this.path}"`
         });
 
         this.s3BucketName = bucketParam.value;
         this.s3ObjectKey = keyParam.value;
 
-        // grant the lambda's role read permissions on the code s3 object
-
         this.bucket = s3.BucketRef.import(parent, 'AssetBucket', {
             bucketName: this.s3BucketName
         });
 
         // form the s3 URL of the object key
         this.s3Url = this.bucket.urlForObject(this.s3ObjectKey);
 
+        // Get a unique identifier for this asset, which will be used
+        // as a folder to group different versions of the same asset together.
+        //
+        // Even though this code looks horrible, this is actually not a terrible thing
+        // to do. The generated ID will contain the *stack name* as well, which is a
+        // perfectly nice thing to do to disambiguate similarly named assets in different stacks.
+        const uniqueId = new cdk.HashedAddressingScheme().allocateAddress(this.path.split('/'));
```

Comment:
I keep wondering if we should just expose a `uniqueId` in the `Construct` level (and use for `resource.logicalId`). This keeps coming back and I think might be useful generally, and not only for CloudFormation resource identifiers.

Reply:
I agree, but their scopes are different, which complicates it a bit.

In a tree like this:

```
root
└── Stack
      └── Construct
             └── Construct

    ├─────── path ───────┤

         ├─ logical id ──┤
```

The `.path` is from the root to the construct, but the logical ID is only calculated off the nearest enclosing stack to the construct. If we didn't do that, all enclosing logical IDs would being with the stack name, which is not terrible for the IDs that they are, but they're also used to generate human-readable names, where that would be quite ugly (and in fact CFN sometimes--but not always!--prepends the stack name itself as well).

So we should probably still keep them separated, or have something like `uniqueIdRelativeTo(ancestor)`

Reply:
We could allow constructs to opt-out of the human portion of name generation, so they will be "skipped" when producing the human portion (something like `hideName = true`).